examples/rust/tracecon/src/bpf/tracecon.bpf.c

```diff
@@ -0,0 +1,168 @@
+	__type(value, struct lookup);
+} lookups SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10240);
+	__type(key, u32);
+	__type(value, struct lookup);
+} hostnames SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10240);
+	__type(key, u32);
+	__type(value, struct sock *);
+} sockets SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+enum tag { IP = 0, HOSTNAME = 1};
+
+struct event {
+	u8 tag;
+	u8 ip[4];
+	u8 hostname[HOSTNAME_LEN];
+};
+
+struct event _event = {0};
+
+static u32 get_tid() {
+	u64 tgid = bpf_get_current_pid_tgid();
+	pid_t pid = (pid_t)(tgid >> 32);
+	if (target_pid != 0 && target_pid != pid)
+		return 0;
+	return (u32)tgid;
+}
+
+SEC("uprobe/getaddrinfo")
+int BPF_KPROBE(getaddrinfo_enter,
+               const char *hostname,
+               const char *service,
+               const struct addrinfo *hints,
+               struct addrinfo **res)
+{
+	u32 tid = get_tid();
+	if (!tid)
+		return 0;
+	struct lookup lookup = {};
+	bpf_probe_read_user_str(&lookup.c, sizeof(lookup.c), hostname);
+	lookup.results = res;
+	bpf_map_update_elem(&lookups, &tid, &lookup, 0);
+	return 0;
+}
+
+SEC("uretprobe/getaddrinfo")
+int BPF_KRETPROBE(getaddrinfo_exit, int ret)
+{
+	u32 tid = get_tid();
+	if (!tid)
+		return 0;
+	struct lookup *lookup = bpf_map_lookup_elem(&lookups, &tid);
+	if (!lookup)
+		return 0;
+	if (ret == 0) {
+		struct addrinfo *result;
+		struct sockaddr *aa;
+		struct in_addr ia;
+		u32 ip;
+		bpf_probe_read_user(&result, sizeof(result), lookup->results);
+		bpf_probe_read_user(&aa, sizeof(aa), &result->ai_addr);
+		bpf_probe_read_user(&ia, 
+		                    sizeof(ia),
+		                    &((struct sockaddr_in *) aa)->sin_addr);
+		ip = ia.s_addr;
+		bpf_map_update_elem(&hostnames, &ip, lookup, 0);
+    }
+	bpf_map_delete_elem(&lookups, &tid);
+	return 0;
+}
+
+SEC("kprobe/tcp_v4_connect")
+int BPF_KPROBE(tcp_v4_connect_enter,
+               struct sock *sk,
+               struct sockaddr *uaddr,
+               int addr_len)
+{
+	u32 tid = get_tid();
+	if (!tid)
+		return 0;
+	bpf_map_update_elem(&sockets, &tid, &sk, 0);
+	return 0;
+};
+
+SEC("kretprobe/tcp_v4_connect")
+int BPF_KRETPROBE(tcp_v4_connect_exit, int ret) {
+	u32 tid = get_tid();
+	if (!tid)
+		return 0;
+	struct sock **sockpp = bpf_map_lookup_elem(&sockets, &tid);
+	if (!sockpp)
+		return 0;
+  	if (ret == 0) {
+		struct event event = {};
+		struct sock *sockp = *sockpp;
+		struct sock_common sock;
+		bpf_probe_read_kernel(&sockp, sizeof(sockp), sockpp);
+		bpf_probe_read_kernel(&sock,
+		                      sizeof(sockp),
+		                      &sockp->__sk_common);
+		u32 daddr = sock.skc_daddr;
+		struct lookup *lookup = bpf_map_lookup_elem(&hostnames, &daddr);
+		if (!lookup) {
+			event.tag = IP;
+			memcpy(&event.ip, &daddr, sizeof(event.ip));
```

Comment:
there is no memcpy() on BPF side, you are just getting like for compiler to inline it for you. Use `__builtin_memcpy()` instead

Reply:
but actually is that such a big deal to have a properly aligned u32 in your event struct?

Reply:
I used __builtin it's aliased w/ a #define on the top. The u32 should be aligned, otherwise we'd have to do it on the rust client code.